Userland/Libraries/LibWebView/Application.cpp

```diff
@@ -5,22 +5,95 @@
  */
 
 #include <AK/Debug.h>
+#include <AK/StringBuilder.h>
+#include <LibCore/ArgsParser.h>
 #include <LibImageDecoderClient/Client.h>
 #include <LibWebView/Application.h>
+#include <LibWebView/URL.h>
 #include <LibWebView/WebContentClient.h>
 
 namespace WebView {
 
 Application* Application::s_the = nullptr;
 
-Application::Application(int, char**)
+Application::Application(
+    Main::Arguments const& arguments,
+    URL::URL new_tab_page_url,
+    Function<void(Core::ArgsParser&)> extra_arguments_creator,
```

Comment:
Instead of these functions, would it make sense to have `virtual parse_arguments(Main::Arguments const&);` and let the chromes override it while subclassing this application object?

Reply:
That gets a bit more awkward than it's worth IMO because
1. we then have to store each extra argument on the class, since they need to be used across both callbacks, and from outside the class
2. We currently do args parsing in the Application constructor. This would have to move to an `init` method that subclasses have to remember to call, because we can't invoke subclass virtual overrides from the base class ctor

Reply:
Ah, in that case I was thinking of usage like so:

Ladybird/<Chome>/Application.h
```
namespace Ladybird {
class Application : public WebView::Application {
    virtual void parse_arguments(..)override;
};
}
```

main.xx
```
Ladybird::Application app;
app.parse_arguments(args);
```

So the responsibility of calling the args parsing routine would move to main.cpp/mm/etc instead of being done in the ctor. And true, the extra arguments and options would need to go on the per-chrome class. but they're already in the per-chrome main.cpp right?

Either way, this is definitely an improvement.

Reply:
Also in theory we could merge the ChromeProcess thingy with the application class later, which would provide an easy out for two-phase init, as that guy has an `static ErrorOr<ChromeProcess> create();` factory

Reply:
Hmm...in an attempt to make each app's `main` have to do as little as possible, maybe something like this would be fine (can't decide if I like it or hate it)

```c++
namespace WebView {
class Application {
public:
    virtual ~Application() = default;

protected:
    template<typename ApplicationType>
    static ApplicationType create(Main::Arguments const& arguments)
    {
        ApplicationType app {};

        Core::ArgsParser args_parser;
        // Do the common set of flags, then:
        app.create_platform_arguments(args_parser);
        args_parser.parse(arguments);

        ChromeOptions chrome_options {};
        WebContentOptions web_content_options {};

        // Do the common set of options, then:
        app.create_platform_options(chrome_options, web_content_options);

        return app;
    }

    Application() = default;

    virtual void create_platform_arguments(Core::ArgsParser&) { }
    virtual void create_platform_options(ChromeOptions&, WebContentOptions&) { }
};

#define WEB_VIEW_APPLICATION                                         \
    friend WebView::Application;                                     \
                                                                     \
public:                                                              \
    static Application create(Main::Arguments const& arguments)      \
    {                                                                \
        return WebView::Application::create<Application>(arguments); \
    }

}

namespace Ladybird {
class Application : public WebView::Application {
    WEB_VIEW_APPLICATION

private:
    Application() = default;

    virtual void create_platform_arguments(Core::ArgsParser&) override
    {
    }

    virtual void create_platform_options(WebView::ChromeOptions&, WebView::WebContentOptions&) override
    {
    }
};
}

ErrorOr<int> serenity_main(Main::Arguments arguments)
{
    auto app = Ladybird::Application::create(arguments);
}
```

Basically want to avoid every main having to create the app instance, call it's parse_args, then call its create_options to create the ChromeOptions / WebViewOptions

> but they're already in the per-chrome main.cpp right?

They are, but just as locals. If they were moved to a class, I'd feel inclined to make them private with const-getters (because they're singleton-y and globally accessible), which then just feels like more boilerplate.

Reply:
aight, I've pushed up a version of this I don't hate 🙃 

Works similarly to what I posted above, but doesn't require that ugly friendship. I've made the following inherit from `WebView::Application`:
1. `Ladybird::Application` in the Qt chrome
2. `Ladybird::ApplicationBridge` in the AppKit chrome (which requires a slight 2-stage init unfortunately, just due to the way we have to create the `NSApplication` in AppKit (we can't do a custom `init`)).
3. A new `Application` struct in headless-browser

Reply:
Nice! I like this version a lot more, and it aligns with what I was expecting to do with the existing Application classes in the chromes.